pkg/querier/frontend2/frontend2.go

```diff
@@ -0,0 +1,290 @@
+func (f *Frontend2) RoundTripGRPC(ctx context.Context, req *httpgrpc.HTTPRequest) (*httpgrpc.HTTPResponse, error) {
+	if s := f.State(); s != services.Running {
+		return nil, fmt.Errorf("frontend not running: %v", s)
+	}
+
+	userID, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Propagate trace context in gRPC too - this will be ignored if using HTTP.
+	tracer, span := opentracing.GlobalTracer(), opentracing.SpanFromContext(ctx)
+	if tracer != nil && span != nil {
+		carrier := (*httpgrpcHeadersCarrier)(req)
+		if err := tracer.Inject(span.Context(), opentracing.HTTPHeaders, carrier); err != nil {
+			return nil, err
+		}
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	freq := &frontendRequest{
+		queryID: f.lastQueryID.Inc(),
+		request: req,
+		userID:  userID,
+
+		cancel: cancel,
+
+		// Buffer of 1 to ensure response or error can be written to the channel
+		// even if this goroutine goes away due to client context cancellation.
+		enqueue:  make(chan enqueueResult, 1),
+		response: make(chan *httpgrpc.HTTPResponse, 1),
+	}
+
+	f.requests.put(freq)
+	defer f.requests.delete(freq.queryID)
+
+	retries := f.cfg.WorkerConcurrency + 1 // To make sure we hit at least two different schedulers.
+
+enqueueAgain:
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+
+	case f.requestsCh <- freq:
+		// Enqueued, let's wait for response.
+	}
+
+	var cancelCh chan<- uint64
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case er := <-freq.enqueue:
+		if er.success {
+			cancelCh = er.cancelCh
+			break
+		}
+
+		if er.retry {
+			retries--
+			if retries > 0 {
+				goto enqueueAgain
+			}
+		}
+
+		return nil, httpgrpc.Errorf(http.StatusInternalServerError, "failed to enqueue request")
+	}
+
+	select {
+	case <-ctx.Done():
+		if cancelCh != nil {
+			select {
+			case cancelCh <- freq.queryID:
+				// cancellation sent.
+			default:
+				// failed to cancel, ignore.
+			}
+		}
+		return nil, ctx.Err()
+
+	case resp := <-freq.response:
+		return resp, nil
+	}
+}
+
+func (f *Frontend2) QueryResult(_ context.Context, qrReq *QueryResultRequest) (*QueryResultResponse, error) {
+	req := f.requests.get(qrReq.QueryID)
+	if req != nil {
+		select {
+		case req.response <- qrReq.HttpResponse:
+			// Should always be possible, unless QueryResult is called multiple times with the same queryID.
+		default:
+			// If we cannot write to the channel, just ignore it.
+		}
+	}
+
+	return &QueryResultResponse{}, nil
+}
+
+// CheckReady determines if the query frontend is ready.  Function parameters/return
+// chosen to match the same method in the ingester
+func (f *Frontend2) CheckReady(_ context.Context) error {
+	workers := f.schedulerWorkers.getWorkersCount()
+
+	// If frontend is connected to at least one scheduler, we are ready.
+	if workers > 0 {
+		return nil
+	}
+
+	msg := fmt.Sprintf("not ready: number of schedulers this worker is connected to is %d", workers)
```

Comment:
You're not missing anything. I had the same comment when this was introduced in original frontend package: https://github.com/cortexproject/cortex/pull/2733#discussion_r447795688

Joe claimed it might be worth just in case it gets negative at some point. I still don't see much value in it, but since it's harmless, I left it as is.